File: Data Structures & Algorithms/climbing-stairs/submission-3.py
class Solution:
    def climbStairs(self, n: int) -> int:
        if(n==2 or n==1):
            return n
        cache = [0 for i in range(n+1)]
        cache[n-1]=1
        cache[n]=1
        for i in range(n-2,-1,-1):
            cache[i] = cache[i+1]+cache[i+2]

        return cache[0]
            

File: Data Structures & Algorithms/coin-change/submission-3.java
class Solution {
    public int coinChange(int[] coins, int amount) {
//        return recursiveHelper(coins, amount, 0);

        int[][] cache = new int[coins.length][amount+1];
        
        return topDownHelper(coins,amount,0,cache);

    }

    private int topDownHelper(int[] coins, int targetAmount, int i, int[][] cache) {
        if (targetAmount == 0) {
            return 0;
        }
        if (i == coins.length) {
            return -1;
        }
        if (cache[i][targetAmount] != 0) {
            return cache[i][targetAmount];
        }

        int ignoreCase = topDownHelper(coins, targetAmount, i+1, cache);
        int keepCase = 0;

        if (targetAmount - coins[i] >= 0) {
            keepCase = topDownHelper(coins, targetAmount - coins[i], i, cache);
            if (keepCase != -1) {
                keepCase++;
            }
        }

        if (ignoreCase * keepCase < 0) {
            cache[i][targetAmount] = Math.max(ignoreCase, keepCase);
        } else {
            cache[i][targetAmount] = Math.min(ignoreCase, keepCase);
        }

        return cache[i][targetAmount];

    }
}
